// @iroha/client/find-api.generated.ts
import * as client from './find-api.generated.prelude.js'
import type * as core from 'jsr:@iroha/core@^0.4.0-beta.1'
import type * as types from 'jsr:@iroha/core@^0.4.0-beta.1/data-model'
export class FindAPI {
  private _executor: client.QueryExecutor
  public constructor(executor: client.QueryExecutor) { this._executor = executor; }
  /** Convenience method for `FindDomains` query, a variant of {@linkcode types.QueryBox} enum. */
  public domains(params?: core.QueryBuilderParams): client.QueryBuilder<'FindDomains'> {
    return new client.QueryBuilder(this._executor, 'FindDomains', params)
  }

  /** Convenience method for `FindAccounts` query, a variant of {@linkcode types.QueryBox} enum. */
  public accounts(params?: core.QueryBuilderParams): client.QueryBuilder<'FindAccounts'> {
    return new client.QueryBuilder(this._executor, 'FindAccounts', params)
  }

  /** Convenience method for `FindAssets` query, a variant of {@linkcode types.QueryBox} enum. */
  public assets(params?: core.QueryBuilderParams): client.QueryBuilder<'FindAssets'> {
    return new client.QueryBuilder(this._executor, 'FindAssets', params)
  }

  /** Convenience method for `FindAssetsDefinitions` query, a variant of {@linkcode types.QueryBox} enum. */
  public assetsDefinitions(params?: core.QueryBuilderParams): client.QueryBuilder<'FindAssetsDefinitions'> {
    return new client.QueryBuilder(this._executor, 'FindAssetsDefinitions', params)
  }

  /** Convenience method for `FindNfts` query, a variant of {@linkcode types.QueryBox} enum. */
  public nfts(params?: core.QueryBuilderParams): client.QueryBuilder<'FindNfts'> {
    return new client.QueryBuilder(this._executor, 'FindNfts', params)
  }

  /** Convenience method for `FindRoles` query, a variant of {@linkcode types.QueryBox} enum. */
  public roles(params?: core.QueryBuilderParams): client.QueryBuilder<'FindRoles'> {
    return new client.QueryBuilder(this._executor, 'FindRoles', params)
  }

  /** Convenience method for `FindRoleIds` query, a variant of {@linkcode types.QueryBox} enum. */
  public roleIds(params?: core.QueryBuilderParams): client.QueryBuilder<'FindRoleIds'> {
    return new client.QueryBuilder(this._executor, 'FindRoleIds', params)
  }

  /** Convenience method for `FindPermissionsByAccountId` query, a variant of {@linkcode types.QueryBox} enum. */
  public permissionsByAccountId(payload: types.FindPermissionsByAccountId, params?: core.QueryBuilderParams): client.QueryBuilder<'FindPermissionsByAccountId'> {
    return new client.QueryBuilder(this._executor, 'FindPermissionsByAccountId', payload, params)
  }

  /** Convenience method for `FindRolesByAccountId` query, a variant of {@linkcode types.QueryBox} enum. */
  public rolesByAccountId(payload: types.FindRolesByAccountId, params?: core.QueryBuilderParams): client.QueryBuilder<'FindRolesByAccountId'> {
    return new client.QueryBuilder(this._executor, 'FindRolesByAccountId', payload, params)
  }

  /** Convenience method for `FindAccountsWithAsset` query, a variant of {@linkcode types.QueryBox} enum. */
  public accountsWithAsset(payload: types.FindAccountsWithAsset, params?: core.QueryBuilderParams): client.QueryBuilder<'FindAccountsWithAsset'> {
    return new client.QueryBuilder(this._executor, 'FindAccountsWithAsset', payload, params)
  }

  /** Convenience method for `FindPeers` query, a variant of {@linkcode types.QueryBox} enum. */
  public peers(params?: core.QueryBuilderParams): client.QueryBuilder<'FindPeers'> {
    return new client.QueryBuilder(this._executor, 'FindPeers', params)
  }

  /** Convenience method for `FindActiveTriggerIds` query, a variant of {@linkcode types.QueryBox} enum. */
  public activeTriggerIds(params?: core.QueryBuilderParams): client.QueryBuilder<'FindActiveTriggerIds'> {
    return new client.QueryBuilder(this._executor, 'FindActiveTriggerIds', params)
  }

  /** Convenience method for `FindTriggers` query, a variant of {@linkcode types.QueryBox} enum. */
  public triggers(params?: core.QueryBuilderParams): client.QueryBuilder<'FindTriggers'> {
    return new client.QueryBuilder(this._executor, 'FindTriggers', params)
  }

  /** Convenience method for `FindTransactions` query, a variant of {@linkcode types.QueryBox} enum. */
  public transactions(params?: core.QueryBuilderParams): client.QueryBuilder<'FindTransactions'> {
    return new client.QueryBuilder(this._executor, 'FindTransactions', params)
  }

  /** Convenience method for `FindBlocks` query, a variant of {@linkcode types.QueryBox} enum. */
  public blocks(params?: core.QueryBuilderParams): client.QueryBuilder<'FindBlocks'> {
    return new client.QueryBuilder(this._executor, 'FindBlocks', params)
  }

  /** Convenience method for `FindBlockHeaders` query, a variant of {@linkcode types.QueryBox} enum. */
  public blockHeaders(params?: core.QueryBuilderParams): client.QueryBuilder<'FindBlockHeaders'> {
    return new client.QueryBuilder(this._executor, 'FindBlockHeaders', params)
  }

  /** Convenience method for `FindExecutorDataModel` query, a variant of {@linkcode types.SingularQueryBox} enum. */
  public executorDataModel(): Promise<core.GetSingularQueryOutput<'FindExecutorDataModel'>> {
    return client.executeSingularQuery(this._executor, 'FindExecutorDataModel')
  }

  /** Convenience method for `FindParameters` query, a variant of {@linkcode types.SingularQueryBox} enum. */
  public parameters(): Promise<core.GetSingularQueryOutput<'FindParameters'>> {
    return client.executeSingularQuery(this._executor, 'FindParameters')
  }

}
