// @iroha/client/util.js
// import type { Debugger } from 'debug'
import Emittery from "emittery";
export function transformProtocolInUrlFromHttpToWs(url) {
  const { protocol } = url;
  if (protocol === 'ws:' || protocol === 'wss:') return url;
  if (protocol === 'http:') {
    const val = new URL(url);
    val.protocol = 'ws:';
    return val;
  }
  if (protocol === 'https:') {
    const val = new URL(url);
    val.protocol = 'wss:';
    return val;
  }
  throw new TypeError(`Expected protocol of ${url} to be on of: ws, wss, http, https`);
}
export function urlJoinPath(url, path) {
  const val = new URL(url);
  if (val.pathname.endsWith('/') && path.startsWith('/')) val.pathname += path.slice(1);
  else val.pathname += path;
  return val;
}
export function setupWebSocket(params) {
  // const debug = params.parentDebugger.extend('websocket')
  const url = urlJoinPath(transformProtocolInUrlFromHttpToWs(params.baseURL), params.endpoint);
  const ee = new Emittery();
  const onceOpened = ee.once('open');
  const onceClosed = ee.once('close');
  // debug('opening connection to %o', url)
  const { isClosed, send, close } = params.adapter.initWebSocket({
    url,
    onopen: (e)=>{
      // debug('connection opened')
      ee.emit('open', e);
    },
    onclose: (e)=>{
      // debug('connection closed; code: %o, reason: %o, was clean: %o', e.code, e.reason, e.wasClean)
      ee.emit('close', e);
    },
    onerror: (e)=>{
      // debug('connection error %o', e)
      ee.emit('error', e);
    },
    onmessage: ({ data })=>{
      // debug('message', data)
      ee.emit('message', data);
    }
  });
  async function closeAsync() {
    if (isClosed()) return;
    // debug('closing connection...')
    close();
    return ee.once('close').then(()=>{});
  }
  async function accepted() {
    return new Promise((resolve, reject)=>{
      onceOpened.then(()=>resolve());
      onceClosed.then(()=>reject(new Error('Handshake acquiring failed - connection closed')));
    });
  }
  return {
    isClosed,
    send,
    close: closeAsync,
    ee,
    accepted
  };
}
